test/Replace.lean

```diff
@@ -0,0 +1,41 @@
+/-
+Copyright (c) 2022 Arthur Paulino. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Arthur Paulino
+-/
+import Mathlib.Tactic.Replace
+
+-- tests with a explicitly named hipothesis
+
+example (h : Int) : Nat := by
+  replace h : Nat := 0
+  exact h
+
+example (h : Nat) : Nat := by
+  have h : Int := 0
+  assumption -- original `h` is not absent but...
+
+example (h : Nat) : Nat := by
+  replace h : Int := 0
+  fail_if_success assumption -- original `h` is absent now
+  replace h : Nat := 0
+  exact h
+
+-- tests with `this`
+
+example : Nat := by
+  have : Int := 0
+  replace : Nat := 0
+  assumption
+
+example : Nat := by
+  have : Nat := 0
+  have : Int := 0
+  assumption -- original `this` is not absent but...
+
+example : Nat := by
+  have : Nat := 0
+  replace : Int := 0
+  fail_if_success assumption -- original `this` is absent now
+  replace : Nat := 0
+  assumption
```

Comment:
Add a test where the old hypothesis can't be cleared because the goal or the new hypothesis depends on it?